File: snapkit-python/snapkit/topology.py
"""
topology.py — SnapTopology: Platonic/ADE Classification
========================================================

The ADE classification is the "periodic table of snap topologies" —
a finite classification of the fundamental shapes that uncertainty
can take. Each ADE type defines a different snap function topology.

"The finiteness is the feature, not the bug. It means the space of
possible constraint topologies is EXPLOREABLE."
— PLATONIC-SNAP-ADE.md
"""

import numpy as np
from dataclasses import dataclass
from typing import Dict, List, Tuple, Optional
from enum import Enum


class ADEType(Enum):
    """ADE root system classification."""
    A1 = "A1"   # Rank 1 (binary)
    A2 = "A2"   # Rank 2 (hexagonal/Eisenstein)
    A3 = "A3"   # Rank 3 (tetrahedral)
    A4 = "A4"   # Rank 4
    D4 = "D4"   # Rank 4 (triality)
    D5 = "D5"   # Rank 5
    E6 = "E6"   # Rank 6 (tetrahedral solid)
    E7 = "E7"   # Rank 7 (octahedral solid)
    E8 = "E8"   # Rank 8 (icosahedral solid — the "noble gas")


# ADE type metadata
ADE_DATA = {
    ADEType.A1: {'rank': 1, 'dim': 2, 'roots': 2, 'coxeter': 2,
                 'solid': None, 'description': 'Binary (coin flip)'},
    ADEType.A2: {'rank': 2, 'dim': 3, 'roots': 6, 'coxeter': 3,
                 'solid': None, 'description': 'Hexagonal (Eisenstein lattice)'},
    ADEType.A3: {'rank': 3, 'dim': 4, 'roots': 12, 'coxeter': 4,
                 'solid': 'Tetrahedron', 'description': 'Tetrahedral (4 categories)'},
    ADEType.A4: {'rank': 4, 'dim': 5, 'roots': 20, 'coxeter': 5,
                 'solid': None, 'description': '5-chain'},
    ADEType.D4: {'rank': 4, 'dim': 4, 'roots': 24, 'coxeter': 6,
                 'solid': None, 'description': 'Triality (D4 symmetry)'},
    ADEType.D5: {'rank': 5, 'dim': 5, 'roots': 40, 'coxeter': 8,
                 'solid': None, 'description': '5-fork'},
    ADEType.E6: {'rank': 6, 'dim': 8, 'roots': 72, 'coxeter': 12,
                 'solid': 'Tetrahedron', 'description': 'Binary tetrahedral group'},
    ADEType.E7: {'rank': 7, 'dim': 8, 'roots': 126, 'coxeter': 18,
                 'solid': 'Cube/Octahedron', 'description': 'Binary octahedral group'},
    ADEType.E8: {'rank': 8, 'dim': 8, 'roots': 240, 'coxeter': 30,
                 'solid': 'Dodecahedron/Icosahedron', 'description': 'Binary icosahedral group'},
}


@dataclass
class SnapTopology:
    """
    The topological structure of a snap function's lattice.
    
    The snap topology determines HOW information is compressed:
    - Hexagonal (A₂): isotropic, 6-fold, densest 2D
    - Cubic (ℤⁿ): axis-aligned, standard grid
    - Tetrahedral (A₃): 4-directional, categorical
    - E₈: maximum symmetry, 8D
    
    The topology is the INVARIANT that transfers across domains.
    When two domains have the same snap topology, calibrated tolerances
    transfer directly.
    """
    ade_type: ADEType
    name: str
    rank: int
    dimension: int
    num_roots: int
    coxeter_number: int
    platonic_solid: Optional[str]
    description: str
    simple_roots: Optional[np.ndarray] = None
    
    @classmethod
    def from_ade(cls, ade_type: ADEType) -> 'SnapTopology':
        """Create a SnapTopology from an ADE type."""
        data = ADE_DATA[ade_type]
        roots = cls._compute_simple_roots(ade_type)
        
        return cls(
            ade_type=ade_type,
            name=ade_type.value,
            rank=data['rank'],
            dimension=data['dim'],
            num_roots=data['roots'],
            coxeter_number=data['coxeter'],
            platonic_solid=data['solid'],
            description=data['description'],
            simple_roots=roots,
        )
    
    @staticmethod
    def _compute_simple_roots(ade_type: ADEType) -> Optional[np.ndarray]:
        """Compute simple root vectors for the given ADE type."""
        if ade_type == ADEType.A1:
            return np.array([[1.0]])
        
        elif ade_type == ADEType.A2:
            # A₂ in 2D: α₁ = (1, 0), α₂ = (-1/2, √3/2)
            return np.array([
                [1.0, 0.0],
                [-0.5, np.sqrt(3)/2],
            ])
        
        elif ade_type == ADEType.A3:
            # A₃ in 3D: standard simple roots
            return np.array([
                [1.0, -1.0, 0.0],
                [0.0, 1.0, -1.0],
                [0.0, 0.0, 1.0],
            ])
        
        elif ade_type == ADEType.D4:
            # D₄ in 4D
            return np.array([
                [1.0, -1.0, 0.0, 0.0],
                [0.0, 1.0, -1.0, 0.0],
                [0.0, 0.0, 1.0, -1.0],
                [0.0, 0.0, 1.0, 1.0],
            ])
        
        # For E6, E7, E8 — use SymPy if available, else skip
        return None
    
    def snap(self, point: np.ndarray) -> Tuple[np.ndarray, float]:
        """
        Snap a point to this topology's root lattice.
        
        Returns (snapped_point, delta_magnitude).
        """
        if self.simple_roots is None:
            raise ValueError(f"Simple roots not computed for {self.name}")
        
        # Project onto root lattice and round
        roots = self.simple_roots
        
        if roots.shape[1] != len(point):
            # Dimension mismatch — pad or truncate
            if roots.shape[1] < len(point):
                padded = np.zeros((roots.shape[0], len(point)))
                padded[:, :roots.shape[1]] = roots
                roots = padded
            else:
                point = np.pad(point, (0, roots.shape[1] - len(point)))
        
        # Solve: point ≈ Σ c_i α_i
        try:
            ATA = roots @ roots.T
            ATp = roots @ point
            coeffs = np.linalg.solve(ATA, ATp)
        except np.linalg.LinAlgError:
            coeffs = np.linalg.lstsq(roots, point, rcond=None)[0]
        
        int_coeffs = np.round(coeffs).astype(int)
        snapped = int_coeffs @ roots
        
        # Handle dimension mismatch
        if len(snapped) > len(point):
            snapped = snapped[:len(point)]
        elif len(snapped) < len(point):
            snapped = np.pad(snapped, (0, len(point) - len(snapped)))
        
        delta = float(np.linalg.norm(point - snapped))
        return snapped, delta
    
    @property
    def quality_score(self) -> float:
        """
        Lattice quality score (higher = better for snap).
        
        Based on: packing density, symmetry, and PID property.
        A₂ has Q ≈ 2.7, E₈ has Q ≈ 3.2 (from our experiments).
        """
        # Approximate quality from root density and symmetry
        return self.num_roots / (self.coxeter_number * self.rank)
    
    def __repr__(self):
        solid = f" → {self.platonic_solid}" if self.platonic_solid else ""
        return (f"SnapTopology({self.name}, rank={self.rank}, "
                f"h={self.coxeter_number}, |Φ|={self.num_roots}{solid})")


# Pre-built topologies for common use cases
def binary_topology() -> SnapTopology:
    """Binary snap — coin flip, true/false, yes/no."""
    return SnapTopology.from_ade(ADEType.A1)

def hexagonal_topology() -> SnapTopology:
    """Hexagonal snap — Eisenstein lattice, densest 2D packing."""
    return SnapTopology.from_ade(ADEType.A2)

def tetrahedral_topology() -> SnapTopology:
    """Tetrahedral snap — 4 categories, categorical decisions."""
    return SnapTopology.from_ade(ADEType.A3)

def triality_topology() -> SnapTopology:
    """D₄ snap — triality symmetry, forked dependencies."""
    return SnapTopology.from_ade(ADEType.D4)

def exceptional_e6() -> SnapTopology:
    """E₆ — tetrahedral solid symmetry."""
    return SnapTopology.from_ade(ADEType.E6)

def exceptional_e7() -> SnapTopology:
    """E₇ — octahedral solid symmetry."""
    return SnapTopology.from_ade(ADEType.E7)

def exceptional_e8() -> SnapTopology:
    """E₈ — icosahedral solid symmetry, maximum finite symmetry."""
    return SnapTopology.from_ade(ADEType.E8)


def all_topologies() -> List[SnapTopology]:
    """Get all pre-built ADE topologies."""
    return [SnapTopology.from_ade(t) for t in ADEType]


def recommend_topology(
    num_categories: Optional[int] = None,
    dimension: Optional[int] = None,
    tensor_rank: Optional[int] = None,
) -> SnapTopology:
    """
    Recommend the best snap topology for given requirements.
    
    Args:
        num_categories: Number of distinct categories/outcomes.
        dimension: Ambient dimension of the data.
        tensor_rank: Required tensor rank for consistent snaps.
    
    Returns:
        The recommended SnapTopology.
    """
    if num_categories == 2:
        return binary_topology()
    elif num_categories == 4:
        return tetrahedral_topology()
    elif dimension == 2:
        return hexagonal_topology()  # A₂ is provably optimal in 2D
    elif tensor_rank is not None and tensor_rank >= 8:
        return exceptional_e8()
    elif tensor_rank is not None and tensor_rank >= 6:
        return exceptional_e7()
    elif tensor_rank is not None and tensor_rank >= 4:
        return triality_topology()
    else:
        return hexagonal_topology()  # Default to A₂ — universal solvent

File: snapkit-js/src/topology.ts
/**
 * Topology — ADE Snap Topology Classification
 *
 * The ADE classification is the "periodic table of snap topologies" —
 * a finite classification of the fundamental shapes that uncertainty
 * can take. Each ADE type defines a different snap function topology.
 *
 * "The finiteness is the feature, not the bug. It means the space of
 * possible constraint topologies is EXPLOREABLE."
 *
 * @module
 */

import type { ADEType, TopologyNode } from './types.js';

// ─── ADE Data Registry ───────────────────────────────────────────────────────

interface ADEData {
  rank: number;
  dimension: number;
  numRoots: number;
  coxeterNumber: number;
  platonicSolid: string | null;
  description: string;
}

const ADE_DATA: Record<ADEType, ADEData> = {
  A1: {
    rank: 1,
    dimension: 2,
    numRoots: 2,
    coxeterNumber: 2,
    platonicSolid: null,
    description: 'Binary (coin flip)',
  },
  A2: {
    rank: 2,
    dimension: 3,
    numRoots: 6,
    coxeterNumber: 3,
    platonicSolid: null,
    description: 'Hexagonal (Eisenstein lattice)',
  },
  A3: {
    rank: 3,
    dimension: 4,
    numRoots: 12,
    coxeterNumber: 4,
    platonicSolid: 'Tetrahedron',
    description: 'Tetrahedral (4 categories)',
  },
  A4: {
    rank: 4,
    dimension: 5,
    numRoots: 20,
    coxeterNumber: 5,
    platonicSolid: null,
    description: '5-chain',
  },
  D4: {
    rank: 4,
    dimension: 4,
    numRoots: 24,
    coxeterNumber: 6,
    platonicSolid: null,
    description: 'Triality (D4 symmetry)',
  },
  D5: {
    rank: 5,
    dimension: 5,
    numRoots: 40,
    coxeterNumber: 8,
    platonicSolid: null,
    description: '5-fork',
  },
  E6: {
    rank: 6,
    dimension: 8,
    numRoots: 72,
    coxeterNumber: 12,
    platonicSolid: 'Tetrahedron',
    description: 'Binary tetrahedral group',
  },
  E7: {
    rank: 7,
    dimension: 8,
    numRoots: 126,
    coxeterNumber: 18,
    platonicSolid: 'Cube/Octahedron',
    description: 'Binary octahedral group',
  },
  E8: {
    rank: 8,
    dimension: 8,
    numRoots: 240,
    coxeterNumber: 30,
    platonicSolid: 'Dodecahedron/Icosahedron',
    description: 'Binary icosahedral group (the "noble gas")',
  },
};

// ─── Simple Root Vectors ─────────────────────────────────────────────────────

/** Get simple root vectors for basic ADE types. Used for lattice projection. */
function computeSimpleRoots(
  adeType: ADEType
): number[][] | null {
  switch (adeType) {
    case 'A1':
      return [[1]];
    case 'A2':
      return [
        [1, 0],
        [-0.5, Math.sqrt(3) / 2],
      ];
    case 'A3':
      return [
        [1, -1, 0],
        [0, 1, -1],
        [0, 0, 1],
      ];
    case 'D4':
      return [
        [1, -1, 0, 0],
        [0, 1, -1, 0],
        [0, 0, 1, -1],
        [0, 0, 1, 1],
      ];
    default:
      return null; // E6/E7/E8 need external libs for full cartan
  }
}

// ─── Main Class ──────────────────────────────────────────────────────────────

/**
 * The topological structure of a snap function's lattice.
 *
 * The snap topology determines HOW information is compressed:
 * - Hexagonal (A₂): isotropic, 6-fold, densest 2D
 * - Cubic (ℤⁿ): axis-aligned, standard grid
 * - Tetrahedral (A₃): 4-directional, categorical
 * - E₈: maximum symmetry, 8D
 *
 * The topology is the INVARIANT that transfers across domains.
 * When two domains have the same snap topology, calibrated tolerances
 * transfer directly.
 */
export class SnapTopology {
  /** ADE type. */
  readonly adeType: ADEType;
  /** Human-readable name (e.g. "A2"). */
  readonly name: string;
  /** Coxeter rank. */
  readonly rank: number;
  /** Ambient dimension. */
  readonly dimension: number;
  /** Number of root vectors. */
  readonly numRoots: number;
  /** Coxeter number (order of Coxeter element). */
  readonly coxeterNumber: number;
  /** Associated Platonic solid, if any. */
  readonly platonicSolid: string | null;
  /** Short description. */
  readonly description: string;
  /** Simple root vectors for lattice projection (may be null for E-types). */
  readonly simpleRoots: number[][] | null;

  constructor(adeType: ADEType) {
    const data = ADE_DATA[adeType];
    this.adeType = adeType;
    this.name = adeType;
    this.rank = data.rank;
    this.dimension = data.dimension;
    this.numRoots = data.numRoots;
    this.coxeterNumber = data.coxeterNumber;
    this.platonicSolid = data.platonicSolid;
    this.description = data.description;
    this.simpleRoots = computeSimpleRoots(adeType);
  }

  /**
   * Create a TopologyNode from this topology.
   */
  toNode(): TopologyNode {
    return {
      adeType: this.adeType,
      name: this.name,
      rank: this.rank,
      dimension: this.dimension,
      numRoots: this.numRoots,
      coxeterNumber: this.coxeterNumber,
      platonicSolid: this.platonicSolid,
      description: this.description,
    };
  }

  /**
   * Snap a point to this topology's root lattice.
   *
   * Returns `[snapped_point, delta_magnitude]`.
   *
   * @param point - The point to snap.
   */
  snap(point: number[]): [number[], number] {
    if (!this.simpleRoots || this.simpleRoots.length === 0) {
      throw new Error(
        `Simple roots not computed for ${this.name}`
      );
    }

    const roots = this.simpleRoots;
    const rank = roots.length;
    const dim = roots[0].length;

    // Handle dimension mismatch by padding/truncating
    let adjustedPoint = [...point];
    if (dim < point.length) {
      adjustedPoint = point.slice(0, dim);
    } else if (dim > point.length) {
      adjustedPoint = [...point, ...new Array(dim - point.length).fill(0)];
    }

    // Project onto root lattice: solve point ≈ Σ c_i α_i
    // Using least-squares: c = (R^T R)^(-1) R^T p
    const RtR: number[][] = Array.from({ length: rank }, () =>
      new Array(rank).fill(0)
    );
    const Rtp: number[] = new Array(rank).fill(0);

    for (let i = 0; i < rank; i++) {
      for (let j = 0; j < rank; j++) {
        RtR[i][j] = roots[i].reduce(
          (sum, v, k) => sum + v * roots[j][k],
          0
        );
      }
      Rtp[i] = roots[i].reduce(
        (sum, v, k) => sum + v * adjustedPoint[k],
        0
      );
    }

    // Solve 2x2 or 3x3 system directly
    const coeffs = this._solveLinear(RtR, Rtp);
    const intCoeffs = coeffs.map(Math.round);

    // Reconstruct snapped point
    const snapped: number[] = new Array(dim).fill(0);
    for (let i = 0; i < rank; i++) {
      for (let j = 0; j < dim; j++) {
        snapped[j] += intCoeffs[i] * roots[i][j];
      }
    }

    // Handle dimension mismatch for output
    let resultSnapped = snapped;
    if (snapped.length > point.length) {
      resultSnapped = snapped.slice(0, point.length);
    } else if (snapped.length < point.length) {
      resultSnapped = [
        ...snapped,
        ...new Array(point.length - snapped.length).fill(0),
      ];
    }

    const delta = Math.sqrt(
      resultSnapped.reduce(
        (sum, v, i) => sum + (point[i] - v) ** 2,
        0
      )
    );

    return [resultSnapped, delta];
  }

  /**
   * Lattice quality score (higher = better for snap).
   *
   * Based on: packing density, symmetry, and PID property.
   * A₂ has Q ≈ 2.7, E₈ has Q ≈ 3.2 (from our experiments).
   */
  get qualityScore(): number {
    return this.numRoots / (this.coxeterNumber * this.rank);
  }

  /**
   * Solve a small linear system Ax = b using Gaussian elimination.
   * Assumes A is square and invertible.
   */
  private _solveLinear(A: number[][], b: number[]): number[] {
    const n = A.length;
    // Augmented matrix
    const aug: number[][] = A.map((row, i) => [...row, b[i]]);

    // Forward elimination
    for (let col = 0; col < n; col++) {
      // Find pivot
      let maxRow = col;
      for (let row = col + 1; row < n; row++) {
        if (Math.abs(aug[row][col]) > Math.abs(aug[maxRow][col])) {
          maxRow = row;
        }
      }
      [aug[col], aug[maxRow]] = [aug[maxRow], aug[col]];

      // Eliminate below
      for (let row = col + 1; row < n; row++) {
        const factor = aug[row][col] / aug[col][col];
        for (let j = col; j <= n; j++) {
          aug[row][j] -= factor * aug[col][j];
        }
      }
    }

    // Back substitution
    const x: number[] = new Array(n).fill(0);
    for (let i = n - 1; i >= 0; i--) {
      x[i] = aug[i][n];
      for (let j = i + 1; j < n; j++) {
        x[i] -= aug[i][j] * x[j];
      }
      x[i] /= aug[i][i];
    }

    return x;
  }
}

// ─── Factory Functions ───────────────────────────────────────────────────────

/** Binary snap — coin flip, true/false, yes/no. */
export function binaryTopology(): SnapTopology {
  return new SnapTopology('A1');
}

/** Hexagonal snap — Eisenstein lattice, densest 2D packing. */
export function hexagonalTopology(): SnapTopology {
  return new SnapTopology('A2');
}

/** Tetrahedral snap — 4 categories, categorical decisions. */
export function tetrahedralTopology(): SnapTopology {
  return new SnapTopology('A3');
}

/** D₄ snap — triality symmetry, forked dependencies. */
export function trialityTopology(): SnapTopology {
  return new SnapTopology('D4');
}

/** E₆ — tetrahedral solid symmetry. */
export function exceptionalE6(): SnapTopology {
  return new SnapTopology('E6');
}

/** E₇ — octahedral solid symmetry. */
export function exceptionalE7(): SnapTopology {
  return new SnapTopology('E7');
}

/** E₈ — icosahedral solid symmetry, maximum finite symmetry. */
export function exceptionalE8(): SnapTopology {
  return new SnapTopology('E8');
}

/** Get all pre-built ADE topologies. */
export function allTopologies(): SnapTopology[] {
  const types: ADEType[] = [
    'A1', 'A2', 'A3', 'A4', 'D4', 'D5', 'E6', 'E7', 'E8',
  ];
  return types.map((t) => new SnapTopology(t));
}

/**
 * Recommend the best snap topology for given requirements.
 *
 * @param options - Requirements for topology selection.
 * @param options.numCategories - Number of distinct categories/outcomes.
 * @param options.dimension - Ambient dimension of the data.
 * @param options.tensorRank - Required tensor rank for consistent snaps.
 */
export function recommendTopology(options: {
  numCategories?: number;
  dimension?: number;
  tensorRank?: number;
}): SnapTopology {
  const { numCategories, dimension, tensorRank } = options;

  if (numCategories === 2) return binaryTopology();
  if (numCategories === 4) return tetrahedralTopology();
  if (dimension === 2) return hexagonalTopology(); // A₂ is provably optimal in 2D
  if (tensorRank !== undefined && tensorRank >= 8)
    return exceptionalE8();
  if (tensorRank !== undefined && tensorRank >= 6)
    return exceptionalE7();
  if (tensorRank !== undefined && tensorRank >= 4)
    return trialityTopology();

  return hexagonalTopology(); // Default to A₂ — universal solvent
}
